### PR TITLE
Optimize lodash. Fixes #6006

### LINKED ISSUE #6006: remove lodash

### PR DESCRIPTION
**Pull request by Omkar76**
### <samp>🤖 Generated by Copilot at 73f49f9</samp>

This pull request replaces the `lodash` dependency with `lodash-es` and imports only the needed functions from the library in various files. This reduces the bundle size and improves the performance of the frontend application. It also refactors some code to use the imported functions directly and consistently.

## Proposed Changes

- Part of #6006 
- Replaced `lodash` with `lodash-es`. 
- Avoid importing entire package as `_`. Only import needed utilities 

@coronasafe/care-fe-code-reviewers @coronasafe/code-reviewers

### <samp>🤖 Generated by Copilot at 73f49f9</samp>

*  Replace `lodash` with `lodash-es` and import only required functions in `package.json` and various files to reduce bundle size and support tree-shaking ([link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L75-R75), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L116-R116), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-6d90523b40f7e247be703cd5d8fccc7e593c6043ade230540bec3295574cff8eL1-R1), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-977c2c274ab7de444d743a9de51baee3ee6e97a077a36211c063af81c42ca841L7-R7), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-40181ebadedb37c4a654bf245c2286ce3cbf9cb836d5262fdb373cb9c5e5032cL1-R1), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-41afad0c2cc9197a2a8f3d9b16519571cc0382b6e790129033b8425c507da0edL20-R20), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-414366305129ef3a3f23768c56e8881943ab3065fcf081bd8ad1986ecb2f5335L1-R10), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-5ea487eaf32006ab39746c942546fa86fa6f4818ff6f812cf877a4fdfec85329L12-R12), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-0bae6f7fb2fdb18df8a369d1cdd7342f00b5716bc1268577f11830c7a3fc2541L4-R4), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-c47b967592a101ec540f7894adb44820da38723774a9a0b8459a05cb54e29fe7L19-R19), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-9e6ecb483e64bf3d105856333925ff3c7a5edf635f9436fb10ca55b25c61b1e8L3-R3), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-1c65fcfe59a39c2fa19c4b333f9dd48e6a8d3c933e82585253b941f8bec98d3eL1-R1), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-09e14a8b77195fb6628b716eb09a71a996a55f559ddc5c3ef19d3c54948a4775L64-R64), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-11b5348acd62c4dc4f1aac97fbeac33804b5c86241edca85599e5ecc57b514b2L10-R10), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-15b7c35aac8026980901950dd4689f22e8fbb3e744340fccd55045f11528d84fL9-R9), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-e2c5c73d556d02503d5adc253828ddf5bad94f6fb9fba25df333d5ca6b9b5c56L3-R3), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-030fed3ca71117f5cbd1dc90f41a4089d207c8fd805a61a960c4e8dda77bbd4eL3-R3))
*  Update usage of imported functions from `lodash-es` to match the import changes in `ExternalResultUpload.tsx`, `Reports/index.tsx`, `Reports/utils.tsx`, `ShowInvestigation.tsx`, `Table.tsx`, `SampleDetails.tsx`, `SampleTestCard.tsx`, and `Notifications.js` ([link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-40181ebadedb37c4a654bf245c2286ce3cbf9cb836d5262fdb373cb9c5e5032cL125-R125), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-41afad0c2cc9197a2a8f3d9b16519571cc0382b6e790129033b8425c507da0edL178-R178), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-414366305129ef3a3f23768c56e8881943ab3065fcf081bd8ad1986ecb2f5335L17-R17), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-414366305129ef3a3f23768c56e8881943ab3065fcf081bd8ad1986ecb2f5335L58-R58), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-5ea487eaf32006ab39746c942546fa86fa6f4818ff6f812cf877a4fdfec85329L113-R113), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-5ea487eaf32006ab39746c942546fa86fa6f4818ff6f812cf877a4fdfec85329L150-R150), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-0bae6f7fb2fdb18df8a369d1cdd7342f00b5716bc1268577f11830c7a3fc2541L62-R62), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-11b5348acd62c4dc4f1aac97fbeac33804b5c86241edca85599e5ecc57b514b2L262-R266), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-11b5348acd62c4dc4f1aac97fbeac33804b5c86241edca85599e5ecc57b514b2L346-R346), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-11b5348acd62c4dc4f1aac97fbeac33804b5c86241edca85599e5ecc57b514b2L429-R429), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-15b7c35aac8026980901950dd4689f22e8fbb3e744340fccd55045f11528d84fL95-R95), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-15b7c35aac8026980901950dd4689f22e8fbb3e744340fccd55045f11528d84fL129-R129), [link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-030fed3ca71117f5cbd1dc90f41a4089d207c8fd805a61a960c4e8dda77bbd4eL46-R46))
*  Remove unused import of `lodash` from `InvestigationTable.tsx` ([link](https://github.com/coronasafe/care_fe/pull/6415/files?diff=unified&w=0#diff-2ae93478d8f7125dcfe12e7bb404ef8476538ea7a08c85fdbb0aa3f42453bcd9L5))

#### Before optimization
lodash_before

#### After  optimization
lodash after

### CODE REVIEW

**Comment by Omkar76** [comment]
cc: @sainak

**Review by rithviknishad — COMMENTED** [review]
@Omkar76 wasn't this issue about removing and using custom utility functions instead?

This PR is just replacing with another package and updating the imports and doesn't make a noticeable difference in the final bundle size.

We could achieve a much lower bundle size if we built the utility functions within care right?

cc: @sainak

**Comment by Omkar76** [comment]
> @Omkar76 wasn't the issue about removing and using custom utility functions instead?
> 
> This PR is just replacing with another package and updating the imports and doesn't make a noticeable difference in the final bundle size.
> 
> We could achieve a much lower bundle size if we built the utility functions within care right?
> 
> cc: @sainak

I believe the size reduction is significant (547 -> 122). We can keep this PR on hold until @sainak has implemented the utilities to replace lodash. If that reduces size significantly more than this we can merge that instead.

**Comment by khavinshankar** [comment]
@rithviknishad @sainak  can you review this pr once, we'll go ahead with this pr for now and update and keep the original issue open to remove this dependency

**Review by rithviknishad — APPROVED** [review]

**Review by sainak — APPROVED** [review]
Lgtm

**Comment by khavinshankar** [comment]
@nihal467 the changes looks good, just give it a general test of above mentioned files

**Comment by nihal467** [comment]
tested

File: src/Components/Patient/SampleTestCard.tsx [hunk 1]
```diff
@@ -126,7 +126,7 @@ export const SampleTestCard = (props: SampleDetailsProps) => {
               Result{" "}
             </div>
             <div className="mt-1 overflow-x-scroll whitespace-normal break-words text-sm font-medium leading-5">
-              {_.startCase(_.camelCase(itemData.result))}
+              {startCase(camelCase(itemData.result))}
```
Comment (bodhish) [inline src/Components/Patient/SampleTestCard.tsx]: Why not use css to achieve the same result; 
https://tailwindcss.com/docs/text-transform#transforming-text

File: src/Components/Patient/SampleTestCard.tsx [hunk 2]
```diff
@@ -92,7 +92,7 @@ export const SampleTestCard = (props: SampleDetailsProps) => {
               Status{" "}
             </div>
             <div className="mt-1 overflow-x-scroll whitespace-normal break-words text-sm font-medium leading-5">
-              {_.startCase(_.camelCase(itemData.status))}
+              {startCase(camelCase(itemData.status))}
```
Comment (bodhish) [inline src/Components/Patient/SampleTestCard.tsx]: Why not use css to achieve the same result; 
https://tailwindcss.com/docs/text-transform#transforming-text

File: src/Components/Patient/SampleDetails.tsx [hunk 3]
```diff
@@ -7,7 +7,7 @@ import ButtonV2 from "../Common/components/ButtonV2";
 import Card from "../../CAREUI/display/Card";
 import { FileUpload } from "./FileUpload";
 import Page from "../Common/components/Page";
-import _ from "lodash";
+import { startCase, camelCase } from "lodash-es";
```
Comment (bodhish) [inline src/Components/Patient/SampleDetails.tsx]: Replace all usages with CSS; 

https://tailwindcss.com/docs/text-transform#transforming-text

File: src/Components/ExternalResult/ExternalResultUpload.tsx [hunk 4]
```diff
@@ -122,7 +122,7 @@ export default function ExternalResultUpload() {
                       ? errors.map((error: any) => {
                           return (
                             <div key={error[0][0]}>
-                              {_.startCase(_.camelCase(error[0][0]))} -{" "}
+                              {startCase(camelCase(error[0][0]))} -{" "}
```
Comment (bodhish) [inline src/Components/ExternalResult/ExternalResultUpload.tsx]: Use CSS; 
https://tailwindcss.com/docs/text-transform#transforming-text

**Review by bodhish — CHANGES_REQUESTED** [review]

**Comment by khavinshankar** [comment]
@Omkar76 make the necessary changes suggested by Bodhi

**Comment by Omkar76** [comment]
Hi @bodhish,
Working on your suggestion to use css text-transform instead of `startCase` and `camelCase`. From what I'm seeing I think it won't be able to replace those methods . Correct me if wrong

> Is a keyword that converts the first letter of each word to uppercase. Other characters remain unchanged (they retain their original case as written in the element's text).
https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform

`sampleDetails.status` can have value "COMPLETED". We need "Completed".  `capitalize` class won't do that. Similar issue for `sampleDetails.doctor_name` and maybe in other places.
https://play.tailwindcss.com/wR1YZEkNsO

**Comment by rithviknishad** [comment]
You could do: 

```jsx
<span className="capitalize">
  {"COMPLETED".toLowerCase()}
</span> 

// 👉 Completed
```

**Review by rithviknishad — APPROVED** [review]

**Review by khavinshankar — APPROVED** [review]

**Pull request merged**